group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorShard.java

```diff
@@ -306,6 +308,19 @@ public CoordinatorResult<OffsetCommitResponseData, Record> commitOffset(
         return offsetMetadataManager.commitOffset(context, request);
     }
 
+    /**
+     * Handles a ListGroups request.
+     *
+     * @param statesFilter The states of the groups we want to list. If empty all groups are returned with their state.
+     * @return A Result containing the ListGroupsResponseData response
+     */
+    public ListGroupsResponseData listGroups(
+        List<String> statesFilter,
+        long committedOffset
+    ) throws ApiException {
+        return new ListGroupsResponseData().setGroups(groupMetadataManager.listGroups(statesFilter, committedOffset));
+    }
```

Comment:
I sill think that we should rather return the list of groups here and create `ListGroupsResponseData` one level up.